src/snowflake/snowpark/_internal/code_generation.py

```diff
@@ -0,0 +1,391 @@
+    complete_source_code = f"""\
+{complete_source_code}
+{f"func = {func_assignment}"}\
+""".strip()
+
+    # 7. if code as comment is true, prefix each line with '#'
+    complete_source_code = (
+        complete_source_code
+        if not code_as_comment
+        else "\n".join(
+            [
+                f"#{f' {line}' if line else ''}"
+                for line in complete_source_code.splitlines()
+            ]
+        )
+    )
+
+    return complete_source_code.strip()
+
+
+def is_lambda(func: FunctionType):
+    return func.__name__ == "<lambda>"
+
+
+def get_lambda_code_text(code_text: str):
+    # add a wrapper to handle the case that the line of lambda source code does not include caller
+    # such that ast could parse the expression tree:
+    #     session.udf.register(
+    #         lambda x, y: x + y, ...
+    #     )
+    try:
+        source_ast = ast.parse(code_text)
+    except SyntaxError as exc:
+        if "cannot assign to lambda" in str(exc):
+            # handle case like:
+            # session.udf.register(
+            #    lambda x, y: x + y, ...
+            # )
+            code_text = f"wrapper({code_text})"
+        elif "unmatched ')'" in str(exc):
+            # handle case like:
+            # session.udf.register(
+            #    lambda x, y: x + y, ...)
+            code_text = f"wrapper({code_text}"
+        source_ast = ast.parse(code_text)
+    lambda_node = next(
+        (node for node in ast.walk(source_ast) if isinstance(node, ast.Lambda)), None
+    )
+    if not lambda_node:
+        raise TypeError("lambda function can not be extracted")
+
+    lines = code_text.splitlines()
+    # single line lambda
+    if len(lines) == 1:
+        return code_text[lambda_node.col_offset : lambda_node.end_col_offset]
+
+    lambda_code_text = ""
+    # lambda of multiple lines
+    # handle case like:
+    # session.udf.register(
+    #    lambda x, y:\
+    #    x + y, ...)
+    for line_idx in range(lambda_node.lineno - 1, lambda_node.end_lineno):
+        line = lines[line_idx]
+        if line_idx == 0:
+            lambda_code_text += f"{line[lambda_node.col_offset:]}\n"
+        elif line_idx == lambda_node.end_lineno - 1:
+            lambda_code_text += line[: lambda_node.end_col_offset]
+        else:
+            lambda_code_text += f"{line}\n"
+    return lambda_code_text
+
+
+def extract_submodule_imports(
+    func: FunctionType, top_level_modules: Iterable[ModuleType]
+):
+    """
+    Get submodule imports, the func code co_names only gives the top level module names, the submodule imports
+    have to be inferred manually. Consider the following example:
+
+    import a1.a2.a3.a4
+    def func():
+        a1.a2.a3.a4.foo()
+
+    func.__code__.co_names only contains ("a1", "a2", "a3", "a4", "foo") which does not include the
+    complete import path information.
+
+    To reconstruct "a1.a2.a3.a4", the current strategy is to import each prefix import of the import chains.
```

Comment:
so it will do 
```
import a1
import a1.a2
import a1.a2.a3
import a1.a2.a3.a4
```
?
Is `co_names` guaranteed ordered?

Reply:
well, yes and no (which is then probably no).

I have configured the following two weird modules for testing:

- a4.a3.a2.a1.SomeClass
- a1.a2.a3.a4.SomeClass

and it turns out `conames` would just come with `(a4, a3, a2, a1, SomeClass)`, so from the behavior, it's like a first-in set.

Reply:
probably it's fine since it's unusual. But can we handle something like this:
```
import pandas.io
import io
```
?
Will it be generated like 
```
import pandas
import pandas.io
```
?

Reply:
I have added test for it, it will be 
```python
import io
import pandas
import pandas.io
```

it's not easy to locate the exact the submodule that is imported. the current implementation imports level by level.

I have created a JIRA for it: https://snowflakecomputing.atlassian.net/browse/SNOW-651634

Reply:
got it